mps_shape_completion_visualization/src/voxel_display.cpp

```diff
@@ -0,0 +1,116 @@
+#include <OGRE/OgreSceneNode.h>
+#include <OGRE/OgreSceneManager.h>
+
+#include <tf/transform_listener.h>
+
+#include <rviz/visualization_manager.h>
+#include <rviz/properties/color_property.h>
+#include <rviz/properties/float_property.h>
+#include <rviz/properties/int_property.h>
+#include <rviz/frame_manager.h>
+
+#include "voxel_visual.h"
+
+#include "voxel_display.h"
+
+namespace mps_shape_completion_visualization
+{
+
+// The constructor must have no arguments, so we can't give the
+// constructor the parameters it needs to fully initialize.
+    VoxelGridDisplay::VoxelGridDisplay()
+    {
+        color_property_ = new rviz::ColorProperty( "Color", QColor( 204, 51, 204 ),
+                                                   "Color to draw the acceleration arrows.",
+                                                   this, SLOT( updateColorAndAlpha() ));
+
+        alpha_property_ = new rviz::FloatProperty( "Alpha Multiple", 1.0,
+                                                   "0 is fully transparent, 1.0 is fully opaque.",
+                                                   this, SLOT( updateColorAndAlpha() ));
+
+
+        binary_display_property_ = new rviz::BoolProperty("Binary Display", true,
+                                                          "If checked, all voxels will have the same alpha",
+                                                          this, SLOT(updateColorAndAlpha() ));
+
+        cutoff_property_ = new rviz::FloatProperty("Threshold", 0.5,
+                                                   "Voxels with values less than this will not be displayed",
+                                                   this, SLOT(updateColorAndAlpha() ));
+
+
+    }
+
+// After the top-level rviz::Display::initialize() does its own setup,
+// it calls the subclass's onInitialize() function.  This is where we
+// instantiate all the workings of the class.  We make sure to also
+// call our immediate super-class's onInitialize() function, since it
+// does important stuff setting up the message filter.
+//
+//  Note that "MFDClass" is a typedef of
+// ``MessageFilterDisplay<message type>``, to save typing that long
+// templated class name every time you need to refer to the
+// superclass.
+    void VoxelGridDisplay::onInitialize()
+    {
+        MFDClass::onInitialize();
+        visual_.reset(new VoxelGridVisual( context_->getSceneManager(), scene_node_ ));
+        updateColorAndAlpha();
+    }
+
+    VoxelGridDisplay::~VoxelGridDisplay()
+    {
+    }
+
+    void VoxelGridDisplay::reset()
+    {
```

Comment:
Does anything need to happen with the `rviz::TypeProperty` pointers here?

Reply:
Changed to unique_ptr, so no.
It isn't so bad to leave the raw pointers anyway